src/preprocessing/passes/non_clausal_simp.cpp

```diff
@@ -448,10 +438,154 @@ PreprocessingPassResult NonClausalSimp::applyInternal(
         Rewriter::rewrite(Node(learnedBuilder)));
   }
 
-  propagator->setNeedsFinish(true);
   return PreprocessingPassResult::NO_CONFLICT;
 }  // namespace passes
 
+SatValue NonClausalSimp::solveByCryptominisat(
+    CVC4::prop::TseitinCnfStream* d_cnfStream,
+    AssertionPipeline* assertionsToPreprocess,
+    unsigned substs_index)
+{
+  for (size_t i = 0, size = assertionsToPreprocess->size(); i < size; ++i)
+  {
+    Trace("non-clausal-simplify") << "Assertion #" << i << " : "
+                                  << (*assertionsToPreprocess)[i] << std::endl;
+  }
+
+  Trace("skeleton-preprocessing") << "asserting to sat solver" << std::endl;
+  SatSolver* d_satSolver = d_cnfStream->getSatSolver();
+  { // for timing
+    TimerStat::CodeTimer timer(d_statistics.d_addingAssertions);
+    for (size_t i = 0, size = assertionsToPreprocess->size(); i < size; ++i)
+    {
+      Assert(Rewriter::rewrite((*assertionsToPreprocess)[i])
+             == (*assertionsToPreprocess)[i]);
+      // Don't reprocess substitutions
+      if (substs_index > 0 && i == substs_index)
+      {
+        continue;
+      }
+      Trace("skeleton-preprocessing")
+          << "asserting " << (*assertionsToPreprocess)[i] << std::endl;
+      d_cnfStream->convertAndAssert(
+          (*assertionsToPreprocess)[i], false, false, RULE_GIVEN);
+    }
+  } // end timing
+
+  return d_satSolver->solve();
+}
+
+std::tuple<bool, std::vector<Node>> NonClausalSimp::preprocessByCryptominisat(
+    AssertionPipeline* assertionsToPreprocess, unsigned substs_index)
+{
+  std::unique_ptr<SatSolver> d_satSolver(SatSolverFactory::createCryptoMinisat(
+      smtStatisticsRegistry(), "non_clausal_simp_solver"));
+  CVC4::prop::NullRegistrar d_registrar;
+  context::Context d_context;
+  CVC4::prop::TseitinCnfStream d_cnfStream(d_satSolver.get(),
+                                           &d_registrar,
+                                           &d_context,
+                                           true,
+                                           "toCNF-skeleton-preprocessing");
+  std::vector<Node> learned_literals;
+  if (solveByCryptominisat(&d_cnfStream, assertionsToPreprocess, substs_index)
+      == SAT_VALUE_FALSE)
+  {
+    // If in conflict, just return false
+    return std::make_tuple(false, learned_literals);
+  }
+
+  std::vector<SatLiteral> topLevelUnits = d_satSolver->getTopLevelUnits();
+  for (size_t i = 0; i < topLevelUnits.size(); i++)
+  {
+    SatLiteral lit = topLevelUnits[i];
+    if (d_cnfStream.getNodeCache().find(lit)
+        != d_cnfStream.getNodeCache().end())
+    {
+      // if the literal is in the getNodeCache
+      Node learnedLiteral = Rewriter::rewrite(d_cnfStream.getNode(lit));
+      if (learnedLiteral.isConst())
```

Comment:
It looks like we are making this distinction already here: https://github.com/CVC4/CVC4/pull/2507/files#diff-8d28c2a099c57ce8aee92960c4ff2db4R144 . Why do it twice?

Reply:
I think in the current way, if we detect a false constant node, we can immediately return.

Reply:
It is not clear to me that this is possible. Assuming that the `learnedLiteral`s are already in rewritten form (which seems to be the case according to your other comment), then if a literal was false, we would assert them as such to the CnfStream and the SAT solver would return `SAT_VALUE_FALSE` if it determined that this literal is actually a unit literal and must be true.

Reply:
Yeah, that makes sense. It's impossible for the SAT solver to have False or Not True in the learned literal.